### PR TITLE
Structify ConfigOptions (#4517)

### LINKED ISSUE #3500: Allow configs to be set with string values
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
In Ballista, I have key-value pairs for configs that I would like to pass along to DataFusion. However, ifI set configs this way and then DataFusion calls `get_u64` or `get_bool`, it will ignore the config because it has the wrong type.

**Describe the solution you'd like**
I would like to be able to pass string for any config and have DataFusion convert to the correct type.

**Describe alternatives you've considered**
None

**Additional context**
None

### LINKED ISSUE #4517: Make ConfigOption names into an Enum
Is there any reason we couldn't just make these an enum? Then we could make the `set_x` and `get_x` methods type safe.

The rationale here is to make the configuration code more "rust like" and follow the Rust way of strongly typed when possible

_Originally posted by @thinkharderdev in https://github.com/apache/arrow-datafusion/pull/4492#discussion_r1039443698_

### LINKED ISSUE #4752: Provide a constructor for the ConfigOptions with HashMap<String, String>
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

It's also related to #3500.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #4517
Closes #3500
Closes #4752
Relates to #3887

# Rationale for this change

The current approach is:

* Not type-safe
* Not documented by rustdoc
* Not understood by code completion
* Expensive to clone
* Defaulting logic is not centralised
* Inconsistently serializes to TaskContext

# What changes are included in this PR?

This reworks ConfigOptions to instead be a struct, using a declarative macro to auto-generate serialization logic

# Are these changes tested?

# Are there any user-facing changes?

Yes

### CODE REVIEW

File: datafusion/core/src/physical_plan/coalesce_batches.rs
Comment (tustvold) [inline datafusion/core/src/physical_plan/coalesce_batches.rs:323]: This will be a lot less peverse once SessionConfig is rolled into ConfigOptions #3887

File: datafusion/core/src/physical_plan/file_format/parquet.rs
Comment (tustvold) [inline datafusion/core/src/physical_plan/file_format/parquet.rs:240]: We no longer have implicit defaults all over the place :smile:

File: datafusion/core/src/config.rs
Comment (tustvold) [inline datafusion/core/src/config.rs:389]: This formulation is specifically designed to be able to replace the AnyMap currently on SessionConfig. As an added bonus it no longer requires types to be interior-mutable

File: datafusion/core/src/config.rs
Comment (tustvold) [inline datafusion/core/src/config.rs:131]: This nesting is a source of non-trivial additional complexity, but I made it work so at least we have support for this in the future.

File: datafusion/core/src/execution/context.rs
Comment (tustvold) [inline datafusion/core/src/execution/context.rs:1384]: This lays the ground-word for eventually removing SessionConfig altogether #3887

File: datafusion/core/src/execution/context.rs
Comment (tustvold) [inline datafusion/core/src/execution/context.rs:1781]: This is a nasty hack, I hope to move `ConfigOptions` into core in a subsequent PR so that it can be used directly instead of via `ContextProvider`

File: datafusion/core/tests/sqllogictests/test_files/information_schema.slt
Comment (tustvold) [inline datafusion/core/tests/sqllogictests/test_files/information_schema.slt:107]: This option was missing from the definitions

File: datafusion/core/tests/sqllogictests/test_files/information_schema.slt
Comment (tustvold) [inline datafusion/core/tests/sqllogictests/test_files/information_schema.slt:108]: `type` is a reserved keyword in rust, and the name was pretty overloaded anyway, so I opted to rename this to `format`. Amusingly this was already what the local variables were named as

File: datafusion/core/tests/fifo.rs
Comment (tustvold) [inline datafusion/core/tests/fifo.rs:91]: This was missed in #4757 the new config returns an error for invalid keys and so caught this

File: datafusion/core/src/execution/context.rs
Comment (tustvold) [inline datafusion/core/src/execution/context.rs:1299]: These two were added in #4757. They're no longer necessary and so I just opted to remove them

File: datafusion/core/src/execution/context.rs
Comment (tustvold) [inline datafusion/core/src/execution/context.rs:1851]: This shows deserializing from a HashMap

File: datafusion/core/src/config.rs
Comment (tustvold) [inline datafusion/core/src/config.rs:329]: This is the mechanism by which `ConfigOptions` can be both serialized and documentation generated

**Review by alamb — COMMENTED** [review]
I really like how the options are now typed / type checked at compile time rather than runtime.

cc @yahoNanJing @andygrove and @thinkharderdev

File: datafusion/core/src/execution/context.rs
Comment (alamb) [inline datafusion/core/src/execution/context.rs:1106]: 👎  for the change to the more overloaded `options` name, even though I realize it is more concise

File: datafusion/core/src/execution/context.rs
Comment (alamb) [inline datafusion/core/src/execution/context.rs:1291]: This may be related to https://github.com/apache/arrow-datafusion/pull/4758 from @yahoNanJing

File: datafusion/core/src/execution/context.rs
Comment (alamb) [reply datafusion/core/src/execution/context.rs:1851]: related to https://github.com/apache/arrow-datafusion/pull/4758

File: docs/source/user-guide/configs.md
Comment (alamb) [inline docs/source/user-guide/configs.md:38]: The major difference here seems to be the type column is no longer present. Was that intended?

File: datafusion/core/src/config.rs
Comment (alamb) [inline datafusion/core/src/config.rs:25]: 👍 

I think some introduction doc comments here would help -- specifically some examples of what this macro is creating would be nice

**Review by alamb — APPROVED** [review]
I love it

The created structure works well in my editor. 

Thank you @tustvold . This is the one missing

It think we should leave this open at least a full work day after approval (aka don't merge until next Monday)

Screenshot 2022-12-30 at 10 13 01 AM

File: datafusion/core/src/execution/context.rs
Comment (tustvold) [reply datafusion/core/src/execution/context.rs:1106]: Yeah, fustfmt was being obtuse and formatting things over 7 lines and this made it happy. SessionConfig isn't long for this world so I think it is fine

File: docs/source/user-guide/configs.md
Comment (tustvold) [reply docs/source/user-guide/configs.md:38]: Yeah, the fields are now Rust typed not datatype. I could include this, but it seemed unnecessary?

Comment (alamb) [reply docs/source/user-guide/configs.md:38]: It is probably clear from the names if something is boolean or numeric. I don't feel strongly

File: datafusion/core/src/physical_optimizer/enforcement.rs
Comment (tustvold) [reply datafusion/core/src/physical_optimizer/enforcement.rs]: Yeah this was just so I could reuse the macro, but the manual implementation isn't so bad, so removed :smile:

**Comment by tustvold** [comment]
I plan to merge this tomorrow morning (GMT) unless anyone needs more time to review

**Pull request merged**
